<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>kr &mdash; pplacer v1.1.alpha19-20-g399f513 documentation</title>
    
    <link rel="stylesheet" href="../_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     'v1.1.alpha19-20-g399f513',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="top" title="pplacer v1.1.alpha19-20-g399f513 documentation" href="../index.html" />
    <link rel="up" title="guppy" href="guppy.html" />
    <link rel="next" title="kr_heat" href="guppy_kr_heat.html" />
    <link rel="prev" title="islands" href="guppy_islands.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="guppy_kr_heat.html" title="kr_heat"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="guppy_islands.html" title="islands"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">pplacer v1.1.alpha19-20-g399f513 documentation</a> &raquo;</li>
          <li><a href="guppy.html" accesskey="U">guppy</a> &raquo;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">kr</a><ul>
<li><a class="reference internal" href="#options">Options</a></li>
<li><a class="reference internal" href="#details">Details</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="guppy_islands.html"
                        title="previous chapter">islands</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="guppy_kr_heat.html"
                        title="next chapter">kr_heat</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/generated_rst/guppy_kr.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="kr">
<span id="guppy-kr"></span><h1>kr<a class="headerlink" href="#kr" title="Permalink to this headline">¶</a></h1>
<p><cite>kr</cite> calculates the Kantorovich-Rubinstein distance and corresponding p-values.</p>
<div class="highlight-python"><div class="highlight"><pre>usage: kr [options] placefiles
</pre></div>
</div>
<div class="section" id="options">
<h2>Options<a class="headerlink" href="#options" title="Permalink to this headline">¶</a></h2>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd><span class="option">--point-mass</span></kbd></td>
<td>Treat every pquery as a point mass concentrated on the highest-weight placement.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--pp</span></kbd></td>
<td>Use posterior probability for the weight.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-c</span></kbd></td>
<td>Reference package path.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-o</span></kbd></td>
<td>Specify the filename to write to.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--out-dir</span></kbd></td>
<td>Specify the directory to write files to.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--prefix</span></kbd></td>
<td>Specify a string to be prepended to filenames.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-p</span></kbd></td>
<td>Exponent for KR integration, i.e. value of p in Z_p. Default 1.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--normalize</span></kbd></td>
<td>Divide KR by a given value. Legal arguments are &#8220;tree-length&#8221;.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--seed</span></kbd></td>
<td>Set the random seed, an integer &gt; 0. Default is 1.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--list-out</span></kbd></td>
<td>Output the KR results as a list rather than a matrix.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--density</span></kbd></td>
<td>Make density plots showing the distribution of randomized values with the calculated values</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-s</span></kbd></td>
<td>Set how many samples to use for significance calculation (0 means calculate distance only). Default is 0.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--gaussian</span></kbd></td>
<td>Use the Gaussian process approximation for p-value estimation</td></tr>
</tbody>
</table>
</div>
<div class="section" id="details">
<h2>Details<a class="headerlink" href="#details" title="Permalink to this headline">¶</a></h2>
<p><tt class="docutils literal"><span class="pre">kr</span></tt> calculates the Kantorovich-Rubinstein distance between collections of placements (given by their place files) by its closed form formula</p>
<div class="math">
<p><img src="../_images/math/b8c3200c42e8bebe1b12dea779a9aa01abbc949d.png" alt="Z(P,Q) =
\int_T \left| P(\tau(y)) - Q(\tau(y)) \right| \, \lambda(dy)."/></p>
</div><p>This is a generalization of the UniFrac distance (UniFrac can only place mass at leaves and cannot accomodate uncertainty).
There is a further generalization to an <img class="math" src="../_images/math/e0b5f1d1ff7f6fd964a048b37a4370d48ea3b9e6.png" alt="L^p"/> Zolotarev-type version:
for <img class="math" src="../_images/math/c9bf86d3757c5b26237be228c2c5d84138927661.png" alt="0 &lt; p &lt; \infty"/> we have the distances</p>
<div class="math">
<p><img src="../_images/math/4cc22f2b11bccf0d92bf189ecadf822f7f71ee66.png" alt="Z_p(P,Q) =
\left[\int_T \left| P(\tau(y)) - Q(\tau(y)) \right|^p \, \lambda(dy)\right]^{\frac{1}{p} \wedge 1}"/></p>
</div><p>which can be used to vary the impact of mass relative to transport.
A larger <img class="math" src="../_images/math/3eca8557203e86160952e1c0f735f7417f3285b1.png" alt="p"/> increases the impact of differences of mass, while a smaller <img class="math" src="../_images/math/3eca8557203e86160952e1c0f735f7417f3285b1.png" alt="p"/> emphasizes distance traveled.</p>
<p>Note that the significance p-values calculated by <tt class="docutils literal"><span class="pre">-s</span></tt> or <tt class="docutils literal"><span class="pre">--gaussian</span></tt> are not corrected for multiple comparison.</p>
<p>The assessment of significance is tricky for metagenomic sampling.
The randomization test (that seems to be very commonly used in association with UniFrac and that is implemented here with the <tt class="docutils literal"><span class="pre">-s</span></tt> flag) does not have wonderful properties when in the setting of incomplete sampling with non-independent observations.
This is commonly the case for metagenomic sampling.
Imagine, for example, that we have a random observation process on the tree equipped with some collection of &#8220;base observations.&#8221;
Each process takes a random subset of those base observations and then throws down some number of reads for each observation in that set, the number of which has mean &gt;&gt; 1.
If the set of base observations is large compared to the number of sample observations, then two draws will always appear significantly different even though they are from the same underlying process.
Thus I would only trust a rejection of the null when sampling is quite deep and the same primers are used for the experiments being compared.</p>
<p>See <a class="reference external" href="http://arxiv.org/abs/1005.1699">Evans and Matsen</a> for more details on phylogenetic KR.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="guppy_kr_heat.html" title="kr_heat"
             >next</a> |</li>
        <li class="right" >
          <a href="guppy_islands.html" title="islands"
             >previous</a> |</li>
        <li><a href="../index.html">pplacer v1.1.alpha19-20-g399f513 documentation</a> &raquo;</li>
          <li><a href="guppy.html" >guppy</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2011-2014, Matsen Group.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.2.
    </div>
  </body>
</html>